src/common/lib/client/resource.ts

```diff
@@ -94,126 +85,240 @@ function urlFromPathAndParams(path: string, params: Record<string, any>) {
   return path + (params ? '?' : '') + paramString(params);
 }
 
-function logResponseHandler<T>(
-  callback: ResourceCallback<T>,
-  method: HttpMethods,
-  path: string,
-  params: Record<string, string>
-): ResourceCallback {
-  return (err, body, headers, unpacked, statusCode) => {
-    if (err) {
-      Logger.logAction(
-        Logger.LOG_MICRO,
-        'Resource.' + method + '()',
-        'Received Error; ' + urlFromPathAndParams(path, params) + '; Error: ' + Utils.inspectError(err)
-      );
-    } else {
-      Logger.logAction(
-        Logger.LOG_MICRO,
-        'Resource.' + method + '()',
-        'Received; ' +
-          urlFromPathAndParams(path, params) +
-          '; Headers: ' +
-          paramString(headers as Record<string, any>) +
-          '; StatusCode: ' +
-          statusCode +
-          '; Body: ' +
-          (Platform.BufferUtils.isBuffer(body) ? body.toString() : body)
-      );
-    }
-    if (callback) {
-      callback(err, body as T, headers, unpacked, statusCode);
-    }
-  };
+function logResult<T>(result: ResourceResult<T>, method: HttpMethods, path: string, params: Record<string, string>) {
+  if (result.err) {
+    Logger.logAction(
+      Logger.LOG_MICRO,
+      'Resource.' + method + '()',
+      'Received Error; ' + urlFromPathAndParams(path, params) + '; Error: ' + Utils.inspectError(result.err)
+    );
+  } else {
+    Logger.logAction(
+      Logger.LOG_MICRO,
+      'Resource.' + method + '()',
+      'Received; ' +
+        urlFromPathAndParams(path, params) +
+        '; Headers: ' +
+        paramString(result.headers as Record<string, any>) +
+        '; StatusCode: ' +
+        result.statusCode +
+        '; Body: ' +
+        (Platform.BufferUtils.isBuffer(result.body) ? result.body.toString() : result.body)
+    );
+  }
 }
 
 export type ResourceCallback<T = unknown> = (
+  /**
+   * Any error returned by the underlying HTTP client.
+   */
   err: IPartialErrorInfo | null,
   body?: T,
   headers?: RequestCallbackHeaders,
   unpacked?: boolean,
   statusCode?: number
 ) => void;
 
+export interface ResourceResponse<T> {
+  body?: T;
+  headers?: RequestCallbackHeaders;
+  unpacked?: boolean;
+  statusCode?: number;
+}
+
+export interface ResourceResult<T> extends ResourceResponse<T> {
+  err: IPartialErrorInfo | null;
+}
+
 class Resource {
-  static get<T = unknown>(
+  /**
+   * @param throwError Whether to throw any error returned by the underlying HTTP client.
+   *
+   * If you specify `true`, then this method will return a `ResourceResponse<T>`, and if the underlying HTTP client returns an error, this method call will throw that error. If you specify `false`, then it will return a `ResourceResult<T>`, whose `err` property contains any error that was returned by the underlying HTTP client.
+   */
+  static async get<T = unknown>(
```

Comment:
unrelated to these changes but: i wonder if it's even worth keeping these helper methods. for Http, Resource, and PaginatedResource it would probably save a bit of bloat if we just used `do(method, ...args)` and got rid of these

Reply:
Sure, added a58e93bc1539dfe808ef0f6a8cea32c98f047f99.